src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DocumentPresentation/AbstractTextDocumentPresentationEndpointBase.cs

```diff
@@ -0,0 +1,362 @@
+               cancellationToken).ConfigureAwait(false);
+
+            if (documentAndVersion is null)
+            {
+                return null;
+            }
+
+            var (documentSnapshot, version) = documentAndVersion;
+            if (documentSnapshot is null)
+            {
+                return null;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                _logger.LogWarning($"Failed to retrieve generated output for document {request.TextDocument.Uri}.");
+                return null;
+            }
+
+            var sourceText = await documentSnapshot.GetTextAsync().ConfigureAwait(false);
+            if (request.Range.Start.TryGetAbsoluteIndex(sourceText, _logger, out var hostDocumentIndex) != true)
+            {
+                return null;
+            }
+
+            var languageKind = _razorDocumentMappingService.GetLanguageKind(codeDocument, hostDocumentIndex);
+            // See if we can handle this directly in Razor. If not, we'll let things flow to the below delegated handling.
+            var result = await TryGetRazorWorkspaceEditAsync(languageKind, request, cancellationToken).ConfigureAwait(false);
+            if (result is not null)
+            {
+                return result;
+            }
+
+            if (languageKind is not (RazorLanguageKind.CSharp or RazorLanguageKind.Html))
+            {
+                _logger.LogInformation($"Unsupported language {languageKind:G}.");
+                return null;
+            }
+
+            var requestParams = CreateRazorRequestParameters(request);
+
+            requestParams.HostDocumentVersion = version;
+            requestParams.Kind = languageKind;
+
+            // For CSharp we need to map the range to the generated document
+            if (languageKind == RazorLanguageKind.CSharp)
+            {
+                if (!_razorDocumentMappingService.TryMapToProjectedDocumentRange(codeDocument, request.Range, out var projectedRange))
+                {
+                    return null;
+                }
+
+                requestParams.Range = projectedRange;
+            }
+
+            var delegatedRequest = await _languageServer.SendRequestAsync(EndpointName, requestParams).ConfigureAwait(false);
+            var response = await delegatedRequest.Returning<WorkspaceEdit?>(cancellationToken).ConfigureAwait(false);
+            if (response is null)
+            {
+                return null;
+            }
+
+            // The responses we get back will be for virtual documents, so we have to map them back to the real
+            // document, and in the case of C#, map the returned ranges too
+            return MapWorkspaceEdit(response, mapRanges: languageKind == RazorLanguageKind.CSharp, codeDocument);
```

Comment:
nit: I'd store the remapped workspace edit in a variable and then return to ease debuggability

Reply:
Yes, you would.